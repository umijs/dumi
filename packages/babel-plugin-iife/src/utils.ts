import type * as BabelCore from '@babel/core';
import * as t from '@babel/types';

export function createDynamicImport(importDeclaration: t.ImportDeclaration) {
  const content = importDeclaration.specifiers.map((specifier) => {
    if (t.isImportNamespaceSpecifier(specifier)) {
      return t.identifier(specifier.local.name);
    }
    if (t.isImportDefaultSpecifier(specifier)) {
      return t.objectProperty(
        t.identifier('default'),
        t.identifier(specifier.local.name),
      );
    }
    const keyName = t.isStringLiteral(specifier.imported)
      ? specifier.imported.value
      : specifier.imported.name;
    return t.objectProperty(
      t.identifier(keyName),
      t.identifier(specifier.local.name),
      false,
      keyName === specifier.local.name,
    );
  });
  let id: t.Identifier | t.ObjectPattern;
  if (t.isIdentifier(content[0])) {
    id = content[0];
  } else {
    id = t.objectPattern(content as t.ObjectProperty[]);
  }
  const vd = t.variableDeclarator(
    id,
    t.awaitExpression(t.callExpression(t.import(), [importDeclaration.source])),
  );
  return t.variableDeclaration('const', [vd]);
}

export function createExportObjectProperty(node: t.ExportDeclaration) {
  if (t.isExportAllDeclaration(node) || !node.declaration) return;

  let { declaration } = node;
  if (
    t.isTSDeclareFunction(declaration) ||
    t.isVariableDeclaration(declaration)
  )
    return;

  let name = 'default';
  if (
    t.isClassDeclaration(declaration) ||
    t.isFunctionDeclaration(declaration)
  ) {
    if (declaration.id && t.isExportNamedDeclaration(node)) {
      name = declaration.id.name;
    }
    declaration = t.toExpression(declaration);
  }
  if (!t.isExpression(declaration)) return;
  return t.objectProperty(t.identifier(name), declaration);
}

export function createIIFE(statements: t.Statement[], async: boolean) {
  return t.expressionStatement(
    t.callExpression(
      t.functionExpression(
        null,
        [],
        t.blockStatement(statements),
        false,
        async,
      ),
      [],
    ),
  );
}

export function isModule(path: BabelCore.NodePath<t.Program>) {
  return path.node.sourceType === 'module';
}
