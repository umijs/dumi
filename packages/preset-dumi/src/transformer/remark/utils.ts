const ATTR_MAPPING = {
  hideactions: 'hideActions',
  defaultshowcode: 'defaultShowCode',
  skipnodemodules: 'skipNodeModules',
  skippropswithoutdoc: 'skipPropsWithoutDoc',
  hidetitle: 'hideTitle',
};

/**
 * parse custome HTML element attributes to properties
 * @note  1. empty attribute will convert to true
 *        2. JSON-like string will convert to JSON
 *        3. workaround for restore property to camlCase that caused by hast-util-raw
 * @param   attrs   original attributes
 * @return  parsed properties
 */
export const parseElmAttrToProps = (attrs: Record<string, string>) => {
  const parsed: Record<string, any> = Object.assign({}, attrs);

  // restore camelCase attrs, because hast-util-raw will transform camlCase to lowercase
  Object.entries(ATTR_MAPPING).forEach(([mark, attr]) => {
    if (parsed[mark] !== undefined) {
      parsed[attr] = parsed[mark];
      delete parsed[mark];
    }
  });

  // convert empty string to boolean
  Object.keys(parsed).forEach(attr => {
    if (parsed[attr] === '') {
      parsed[attr] = true;
    }
  });

  // try to parse JSON field value
  Object.keys(parsed).forEach(attr => {
    if (/^(\[|{)[^]*(]|})$/.test(parsed[attr])) {
      try {
        parsed[attr] = JSON.parse(parsed[attr]);
      } catch (err) {
        /* nothing */
      }
    }
  });

  return parsed;
};
