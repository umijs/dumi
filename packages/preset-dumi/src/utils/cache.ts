import fs from 'fs';

export default class FileCache {
  cache: Record<string, { filePath: string; updatedTime: number; value: any }> = {};

  add(filePath: string, value: any, key?: string) {
    this.cache[key || filePath] = {
      filePath,
      value,
      updatedTime: fs.lstatSync(filePath).mtimeMs,
    };
  }

  get(key: string) {
    let result;

    if (
      this.cache[key] &&
      fs.lstatSync(this.cache[key].filePath).mtimeMs === this.cache[key].updatedTime
    ) {
      result = this.cache[key].value;
    }

    return result;
  }
}
