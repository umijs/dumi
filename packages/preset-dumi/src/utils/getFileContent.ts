import ctx from '../context';

/**
 * get range lines of markdown file
 */
export const getFileRangeLines = (content: string, range: string) => {
  const [, start, end] = range?.match(/^L(\d+)(?:-L(\d+))?$/) || [];

  if (start) {
    const lineStart = parseInt(start, 10) - 1;
    const lineEnd = end ? parseInt(end, 10) : lineStart + 1;

    content = content
      .split(/\r\n|\n/g)
      .slice(lineStart, lineEnd)
      .join('\n');
  }

  return content;
};


/**
 * get file content by regular expression
 * @param content   source file content
 * @param regexp    regular expression string
 * @param filePath  source file path
 */
export const getFileContentByRegExp = (content: string, regexp: string, filePath: string) => {
  try {
    // eslint-disable-next-line no-eval
    return content.match(eval(regexp))[0];
  } catch (err) {
    ctx.umi?.logger.error(`[dumi]: extract content failed, use the full content.
RegExp: ${regexp}
File: ${filePath}
Error: ${err}`);
    return content;
  }
}
