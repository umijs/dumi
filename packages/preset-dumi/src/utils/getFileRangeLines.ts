/**
 * get range lines of markdown file
 */
export default (content: string, range: string) => {
  const [, start, end] = range?.match(/^L(\d+)(?:-L(\d+))?$/) || [];

  if (start) {
    const lineStart = parseInt(start, 10) - 1;
    const lineEnd = end ? parseInt(end, 10) : lineStart + 1;

    content = content
      .split(/\r|\n/g)
      .slice(lineStart, lineEnd)
      .join('\n');
  }

  return content;
};
