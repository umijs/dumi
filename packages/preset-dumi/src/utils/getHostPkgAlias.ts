import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';
import type { IApi } from '@umijs/types';

function getPkgAliasForPath(absPath: string) {
  const result: [string, string] = ['', absPath];
  const pkgPath = path.join(absPath, 'package.json');

  // use package.name if exists
  if (fs.existsSync(pkgPath)) {
    result[0] = require(pkgPath).name;
  }

  return result;
}

export default (paths: IApi['paths']) => {
  const isLerna = fs.existsSync(path.join(paths.cwd, 'lerna.json'));
  const isPnpmWorkspace = fs.existsSync(path.join(paths.cwd, 'pnpm-workspace.yaml'))
  const pkgs: [string, string][] = [];

  if (isLerna) {
    // for lerna repo
    const { version: lernaVersion } = require('lerna/package.json');
    const lernaMainVersion = Number(lernaVersion.split('.')[0]);
    if (lernaMainVersion >= 3) {
      JSON.parse(
        execSync(`${path.join(paths.cwd, 'node_modules/.bin/lerna')} ls --json --all`, {
          stdio: 'pipe',
          // fix: 修复windows环境下有多余输出导致JSON.parse报错的问题
        })
          .toString()
          .replace(/([\r\n]\])[^]*$/, '$1'),
      ).forEach(pkg => {
        pkgs.push([pkg.name, pkg.location]);
      });
    } else if (require.resolve('lerna/lib/PackageUtilities', { paths: [paths.cwd] })) {
      // reference: https://github.com/azz/lerna-get-packages/blob/master/index.js
      const PackageUtilities = require(require.resolve('lerna/lib/PackageUtilities', {
        paths: [paths.cwd],
      }));
      const Repository = require(require.resolve('lerna/lib/Repository', { paths: [paths.cwd] }));

      PackageUtilities.getPackages(new Repository(paths.cwd)).forEach(pkg => {
        pkgs.push([pkg._package.name, pkg._location]);
      });
    }
  } else if (isPnpmWorkspace) {
    JSON.parse(execSync('pnpm list -r --json', {
          stdio: 'pipe',
        })
          .toString()
          .replace(/([\r\n]\])[^]*$/, '$1')
    ).filter(pkg => pkg.path !== paths.cwd) // filter root pkg
    .forEach(pkg => {
        pkgs.push([pkg.name, pkg.path]);
      });
  } else {
    // for standard repo
    pkgs.push(getPkgAliasForPath(paths.cwd));
  }

  return pkgs;
};
