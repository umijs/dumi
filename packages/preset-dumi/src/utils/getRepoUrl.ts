import hostedGit from 'hosted-git-info';

export default (url: any, platform?: 'gitlab') => {
  if (!url || typeof(url) !== 'string') return '';

  let repoUrl = hostedGit.fromUrl(url)?.browse();

  if (!repoUrl) {
    const isHttpProtocol = url.includes('http://');

    if (platform === 'gitlab') {
      if (isHttpProtocol) url = url.replace('http', 'https');

      let originalHost: string;

      repoUrl = hostedGit
        .fromUrl(
          // fake to gitlab to make hostedGit worked
          // refer: https://github.com/npm/hosted-git-info/pull/30#issuecomment-400074956
          url.replace(/([\w-]+\.)+[\w-]+/, str => {
            originalHost = str;

            return 'gitlab.com';
          }),
        )
        ?.browse()
        // restore the original host
        ?.replace('gitlab.com', originalHost);
    }

    // process other case, protocol://domain/group/repo{discard remaining paths}
    repoUrl =
      repoUrl || url.replace(/^.*?((?:[\w-]+\.?)+)+[:/]([\w-]+)\/([\w-]+).*$/, 'https://$1/$2/$3');

    if (isHttpProtocol) repoUrl = repoUrl.replace('https', 'http');
  }

  return repoUrl;
};
