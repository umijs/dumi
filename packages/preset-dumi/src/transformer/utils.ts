import fs from 'fs';
import { winPath } from '@umijs/utils';
import ctx from '../context';

/* eslint-disable no-new-wrappers */
/**
 * transform props base on JSX rule
 * @param props   original props
 */
export const formatJSXProps = (props: Record<string, any>): Record<string, any> => {
  const OMIT_NULL_PROPS = ['alt', 'align'];

  return Object.keys(props || {}).reduce((result, key) => {
    // ignore useless empty props
    if (props[key] !== null || !OMIT_NULL_PROPS.includes(key)) {
      result[key] = props[key];
    }

    // use wrapper object for workaround implement raw props render
    // https://github.com/mapbox/jsxtreme-markdown/blob/main/packages/hast-util-to-jsx/index.js#L167
    if (
      !(props[key] instanceof String) &&
      props[key] !== null &&
      (typeof props[key] === 'object' || Array.isArray(props[key]))
    ) {
      result[key] = new String(JSON.stringify(props[key]));
    }

    // join className to string
    if (key === 'className' && Array.isArray(props[key])) {
      result[key] = props[key].join(' ');
    }

    return result;
  }, {});
};

/**
 * get umi dynamicImport flag
 */
export function isDynamicEnable() {
  return Boolean(ctx.umi?.config?.dynamicImport);
}

/**
 * encode & decode import statement which need to hoist at the top of file
 */
const HOIST_ID = '^H^O^I^S^T^';
export const encodeHoistImport = (resolvePath: string) =>
  `import ${HOIST_ID} from '!!dumi-raw-code-loader!${winPath(resolvePath)}?dumi-raw-code'`;
export const decodeHoistImport = (str: string, id: string) =>
  str.replace(new RegExp(HOIST_ID.replace(/\^/g, '\\^'), 'g'), id);
export const isHoistImport = (str: string) => str.startsWith(`import ${HOIST_ID} from`);
export const decodeHoistImportToContent = (str: string) => {
  if (isHoistImport(str)) {
    const filePath = str.match(/dumi\-raw\-code\-loader!([^\?]+)\?/)?.[1];

    return fs.readFileSync(filePath, 'utf-8').toString();
  }

  return str;
};

/**
 * encode import/require statement by dynamicImport, it can be decode to await import statement with chunkName
 */
export const CHUNK_ID = '^C^H^U^N^K^';
export const encodeImportRequire = (resolvePath: string) =>
  `(${isDynamicEnable() ? `await import(${CHUNK_ID}` : 'require('}'${winPath(
    resolvePath,
  )}')).default`;
export const decodeImportRequire = (str: string, chunkName: string) =>
  str.replace(
    new RegExp(CHUNK_ID.replace(/\^/g, '\\^'), 'g'),
    `/* webpackChunkName: "${chunkName}" */`,
  );
export const isEncodeImport = (str: string) => str.startsWith(`(await import(${CHUNK_ID}`);
export const decodeImportRequireWithAutoDynamic = (str: string, chunkName: string) =>
  isEncodeImport(str)
    ? `dynamic({
      loader: async () => ${decodeImportRequire(str, chunkName)},
    })`
    : decodeImportRequire(str, chunkName);
