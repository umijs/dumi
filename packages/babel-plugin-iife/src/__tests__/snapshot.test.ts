import { describe, expect, test } from 'vitest';
import { transpile } from './utils';

describe('babel-plugin-iife', () => {
  [
    {
      name: 'should wrapped by iife',
      from: 'const a = 1;',
    },
    {
      name: 'static import -> dynamic import',
      from: `
        import a from 'a';
        import { b } from 'b';
        import { c1 as c } from 'c';
        import * as d from 'd';
        import e, { e1, e2 as e3 } from 'e';
        import { createVNode as _createVNode, createTextVNode as _createTextVNode } from "vue";
      `,
    },
    {
      name: 'export default `value`',
      from: 'export default a;',
    },
    {
      name: 'export default `class`',
      from: 'export default class A {};',
    },
    {
      name: 'export default `function`',
      from: 'export default () => null;',
    },
    {
      name: 'export named',
      from: `
        const name1 = 1, name2 = 2;
        export { name1, name2 };
        export let name3 = 1;
        export function x () {};
        export class A {};
      `,
    },
  ].forEach(({ name, from }) => {
    test(name, async () => {
      expect(await transpile(from)).toMatchSnapshot();
    });
  });
});
