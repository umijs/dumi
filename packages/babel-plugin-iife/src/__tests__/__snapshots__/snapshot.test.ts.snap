// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`babel-plugin-iife > export default \`class\` 1`] = `
"(function () {
  ;
  return {
    default: class A {}
  };
})();"
`;

exports[`babel-plugin-iife > export default \`function\` 1`] = `
"(function () {
  return {
    default: () => null
  };
})();"
`;

exports[`babel-plugin-iife > export default \`value\` 1`] = `
"(function () {
  return {
    default: a
  };
})();"
`;

exports[`babel-plugin-iife > export named 1`] = `
"(function () {
  const name1 = 1,
    name2 = 2;
  ;
  ;
  return {
    x: function x() {},
    A: class A {}
  };
})();"
`;

exports[`babel-plugin-iife > should wrapped by iife 1`] = `
"(function () {
  const a = 1;
})();"
`;

exports[`babel-plugin-iife > static import -> dynamic import 1`] = `
"(async function () {
  const {
    default: a
  } = await import('a');
  const {
    b
  } = await import('b');
  const {
    c1: c
  } = await import('c');
  const d = await import('d');
  const {
    default: e,
    e1,
    e2: e3
  } = await import('e');
  const {
    createVNode: _createVNode,
    createTextVNode: _createTextVNode
  } = await import(\\"vue\\");
})();"
`;
